awx/api/serializers.py

```diff
@@ -1388,6 +1388,11 @@ def get_field_from_model_or_attrs(fd):
         elif self.instance:
             organization = self.instance.organization
 
+        if 'allow_override' in attrs and self.instance:
+            if attrs['allow_override'] != self.instance:
+                raise serializers.ValidationError({
+                    'allow_override': _('Branch override behavior of a project cannot be changed after creation.')})
```

Comment:
Why? Even if we know why we need to be able to tell people why.

Reply:
1) Not sure I like this
2) dont' show me button if I can't do it
3) This is not coming up through UI, getting mysterious 400 instead:
![Screenshot from 2019-07-24 14-05-29](https://user-images.githubusercontent.com/29239279/61817522-c3466a80-ae1c-11e9-92c4-0c569ccacd55.png)

Reply:
Let me recap the reason that this was added, say someone does this:

 - make project with `allow_override` being true
 - makes JT with a non-blank `scm_branch`
 - changes project `allow_override` field back to False

Now the user has a JT that specifies a branch that we cannot apply.

Providing an outline of this case is far too much wording to put in an error message, but I agree the UI should at least show the error message from the API. Or hide it when in edit mode.

Also, this is more complicated than the other "prompts" related rejections of fields. If a JT's field is rejected because the project is configured to not allow it, we currently have no mechanism or precedent for dealing with that. We have no way of letting the user know that's broken without a lot of novel UI work.

Reply:
After more discussions, I agree the API side of this can be made smoother for the user interaction.

It would be fairly easy to combine a check with this, something like `JobTemplate.objects.filter(project=obj).exclude(scm_branch="").exists()`, and then if both this condition is satisfied, and the `allow_override` value changes from its prior value, then we could give a more specific error message:

> Branch override behavior of this project is currently relied on by one or more existing job templates.

Then if there are no such JTs, we can just allow the PATCH/PUT.

Reply:
+1 to @kdelee's view of things and your solution seems to be a step in the right direction @AlanCoding , it seems allowing the change makes more sense for the end user -- it would also make it easier if the dependent templates could be listed, though I'm unsure of performance impact.

Reply:
> it would also make it easier if the dependent templates could be listed

That's possible, but depends on how.

If all you're suggesting is a list of names or primary keys in the error message, then sure. That is super easy, and IMO no performance impact that matters. But this will not be linked.

Giving primary keys would be the most concise. If the message is too long, it will overflow the popup box, because error messages are intended to be concise. But if I give primary keys, I worry that someone will say that's a bad UX. But I think handling any special UI cases here is going way overboard for a corner case.

Reply:
@AlanCoding I see the issue. How about something like a warning that existing job templates use branch override behavior, and offer a confirm button that would revert them if the person really doesn't care?